modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java

```diff
@@ -0,0 +1,432 @@
+      return displayName;
+    }
+    List<String> givenNames = attributeMapper.getMappedAttributes(request, "givenName");
+    String givenName = "";
+    if (givenNames.size() > 0) {
+      String givenNameValue = givenNames.get(0);
+      givenName = StringUtils.isBlank(givenNameValue) ? ""
+              : new String(givenNameValue.getBytes(StandardCharsets.ISO_8859_1),
+                      StandardCharsets.UTF_8);
+    }
+    List<String> surnames = attributeMapper.getMappedAttributes(request, "sn");
+    String sn = "";
+    if (surnames.size() > 0) {
+      String surnameValue = surnames.get(0);
+      sn = StringUtils.isBlank(surnameValue) ? ""
+              : new String(surnameValue.getBytes(StandardCharsets.ISO_8859_1),
+                      StandardCharsets.UTF_8);
+    }
+    return StringUtils.join(new String[] { givenName, sn }, " ");
+  }
+
+  /**
+   * Extracts the e-mail from the request.
+   *
+   * @param request
+   *          the request
+   * @return the e-mail address
+   */
+  private String extractEmail(HttpServletRequest request) {
+    List<String> mailAdresses = attributeMapper.getMappedAttributes(request, "mail");
+    if (mailAdresses.size() > 0) {
+      String mailValue = mailAdresses.get(0);
+      String mail = StringUtils.isBlank(mailValue) ? ""
+              : new String(mailValue.getBytes(StandardCharsets.ISO_8859_1),
+                      StandardCharsets.UTF_8);
+      return mail;
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Extracts the e-mail from the request.
+   *
+   * @param request
+   *          the request
+   * @return the e-mail address
+   */
+  private String extractDisplayName(HttpServletRequest request) {
+    List<String> displayNames = attributeMapper.getMappedAttributes(request, "displayName");
+    if (displayNames.size() > 0) {
+      String displayNameValue = displayNames.get(0);
+      String displayName = StringUtils.isBlank(displayNameValue) ? ""
+              : new String(displayNameValue.getBytes(StandardCharsets.ISO_8859_1),
+                      StandardCharsets.UTF_8);
+      return displayName;
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Extracts the roles from the request.
+   *
+   * @param request
+   *          the request
+   * @return the roles
+   */
+  private Set<JpaRole> extractRoles(String id, HttpServletRequest request) {
+    List<String> aaiRoles = attributeMapper.getMappedAttributes(request, "roles");
+    JpaOrganization organization = fromOrganization(securityService.getOrganization());
+    Set<JpaRole> roles = new HashSet<JpaRole>();
+    if (aaiRoles != null) {
+      for (String aaiRole : aaiRoles) {
+        roles.add(new JpaRole(aaiRole, organization));
+      }
+    }
+    roles.add(new JpaRole(organization.getAnonymousRole(), organization));
+    if (StringUtils.equals(id, bootstrapUserId)) {
+      roles.add(new JpaRole(GLOBAL_ADMIN_ROLE, organization));
+    }
```

Comment:
What's the rule for assigning the global admin role here?

Reply:
Hi @lkiesow, long time no see... As I originally wrote this code, I can give some comments here.

To answer your question: https://github.com/opencast/opencast/blob/develop/etc/org.opencastproject.security.aai.ConfigurableLoginHandler.cfg#L12

@jchssystems Assuming that the DynamicLoginHanlder replaces the ConfigurableLoginHandler, the bootstrap user logic is obsolete as you can configure a bootstrap user directly in the security configuration.
But the following changes seem missing to me:
- Delete ConfigurableLoginHandler.java
- Delete the configuration file of the ConfigurableLoginHandler (https://github.com/opencast/opencast/blob/develop/etc/org.opencastproject.security.aai.ConfigurableLoginHandler.cfg)
- Properly adjust the documentation (https://github.com/opencast/opencast/blob/develop/docs/guides/admin/docs/configuration/security.aai.md). This should include information about how bootstrapping works (even if the example in the security configuration makes this obvious to me)

@JulianKniephoff This change should be mentioned in https://docs.opencast.org/develop/admin/#upgrade/#_top (Upgrading 8.x to 9.x) as it is not obvious that adaptors which use Shibboleth need to adapt their configuration quite a bit.

Reply:
@staubesv good to hear from you again, and thanks for the heads up. :smile: Will do when/if this is merged.

Reply:
> What's the rule for assigning the global admin role here?

Backward compatibility. However it is also possible to assign this ROLE in the SpEL configuration, so we will remove `bootstrapUserId`

Reply:
> * Delete ConfigurableLoginHandler.java
> 
> * Delete the configuration file of the ConfigurableLoginHandler (https://github.com/opencast/opencast/blob/develop/etc/org.opencastproject.security.aai.ConfigurableLoginHandler.cfg)

It is still possible to use the old ConfigurableLoginHandler. Our intention was a new / additional LoginHandler, not to replace the old one.

Reply:
@jchssystems As far as I understand, the DynamicLoginHandler offers a superset of the functionality of the ConfigureableLoginHandler. Why should Opencast include both the ConfigurableLoginHandler and the DynamicLoginHandler? That is just more code with no added value...

Reply:
@staubesv we did not want to break existing configurations for backward compatibility, not every adopter needs this functionality. My proposal: we continue with both handlers, first writing a dedicated documentation for the DynamicLoginHandler, fixing all issues and finally make a new PR to remove the ConfigurableLoginHandler

Reply:
@jchssystems No strong opinion, but you could expect adopters to change the configuration. Note, however, that "fixing all issues" should be done before merging ;-)

Reply:
bootstrapUserId was removed.
A new independent documentation will be made when all the code be fixed.

Reply:
In my opinion you should not force to use one or another login handler (or a third in the future).